p2pserver/message/utils/msg_workerpool.go

```diff
@@ -0,0 +1,225 @@
+/*
+ * Copyright (C) 2018 The ontology Authors
+ * This file is part of The ontology library.
+ *
+ * The ontology is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Lesser General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * The ontology is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with The ontology.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package utils
+
+import (
+	"runtime"
+	"sync"
+	"time"
+
+	"github.com/ontio/ontology-eventbus/actor"
+	"github.com/ontio/ontology/common/log"
+	"github.com/ontio/ontology/p2pserver/message/types"
+	"github.com/ontio/ontology/p2pserver/net/protocol"
+)
+
+type msgJobChanList *[]*msgJobChan
+
+type msgWorkerPool struct {
+	maxWorkerCount    uint
+	maxWorkerIdleTime time.Duration
+	curWorkerCount    uint
+	stopChan          chan struct{}
+	waitingWokers     map[string]msgJobChanList
+	workChanPool      sync.Pool
+	lock              sync.Mutex
+}
+
+type msgJobItem struct {
+	msgPayload *types.MsgPayload
+	p2p        p2p.P2P
+	pid        *actor.PID
+	msgHandler MessageHandler
+}
+
+type msgJobChan struct {
+	lastUseTime time.Time
+	jobChan     chan *msgJobItem
+}
+
+func (this *msgWorkerPool) init() {
+
+	this.waitingWokers = make(map[string]msgJobChanList)
+	this.stopChan = make(chan struct{})
+	this.maxWorkerIdleTime = 10 * time.Second
+}
+
+var msgJobChanCap = func() int {
+	// Use blocking msgJobChan if GOMAXPROCS=1.
+	// This immediately switches receiveMsg to disposeMsg, which results
+	// in higher performance
+	if runtime.GOMAXPROCS(0) == 1 {
+		return 0
+	}
+
+	// Use non-blocking msgJobChan if GOMAXPROCS>1,
+	// since otherwise the receiveMsg caller (Acceptor) may lag accepting
+	// new message if disposeMsg is CPU-bound.
+	return 1
+}()
+
+func (this *msgWorkerPool) start() {
+
+	if this.waitingWokers == nil || this.stopChan == nil {
+		log.Error("[p2p]invalid start invoking, the msg worker pool hasn't been initialized")
+		return
+	}
+
+	go func() {
+		for {
+			this.clean()
```

Comment:
to be simple, why not start all workers here, and let them waiting on one channel?

Reply:
If the peer connection count is smaller, if all workers are waiting,  the resources are waste.